### ISSUE TITLE
How to deal with vectorised integration limits

### ISSUE THREAD

**Issue by HJA24** [saullocastro/cubature#34]
I am struggling with using `cubature.cubature` when `xmin` and `xmax` are not scalar values.

My code looks as follows:

```
from cubature import cubature
import numpy as np
import scipy.stats as ss


class Kelly:
    def __init__(self):
        self.theo = 0.556
        self.odds = 1.952
        self.kelly = self.calculate_kelly()
        self.x = 100
        self.n = 180
        self.a = 50
        self.b = 50
        self.i = 0.001
        self.f = np.arange(0, 1 + self.i, self.i)
        self.c1 = 1
        self.c2 = 2
        self.k = 1.5

    def calculate_kelly(self) -> float:
        return (self.theo * self.odds - 1) / (self.odds - 1)

    def calculate_loss_function_vec(self, p):
        f = self.f.flatten()
        p = p[:, 0]
        loss_function = np.where(p[:, None] < (f * self.odds - 1) + 1 / self.odds,
                                 (self.c1 + self.c2) * abs(f - self.kelly) ** self.k, 0)
        return loss_function * ss.beta.pdf(p, self.a + self.x, self.n - self.x + self.b)[:, None]

    def integrate_vec(self):
        vals, errors = cubature(func=self.calculate_loss_function_vec,
                                ndim=1,
                                fdim=len(self.f),
                                xmin=np.zeros((1, len(self.f))),
                                xmax=np.array([self.f * (self.odds - 1) + 1 / self.odds]),
                                vectorized=True)
        return vals, errors


kelly = Kelly()
vals, errors = kelly.integrate_vec()
print(vals, errors)
```

My code is based on the code in the Appendix of this [paper](https://www.degruyter.com/document/doi/10.1515/jqas-2017-0122/html). In this paper they use the R-wrapper of cubature

![image](https://user-images.githubusercontent.com/30417550/188444131-31a9f7ec-f65a-4b20-bfcf-44c4a380280b.png)

Here they specify `xmin` as 0 and `xmax` as a one-row vector with 1001 columns.

Thanks for looking into it!

I am using Python 3.7 and cubature 0.15.0

**Comment by saullocastro** [comment]
Sorry for the late reply @HJA24 . I'm looking into it.

**Comment by saullocastro** [comment]
@HJA24  I think I could understand your question. The vectorized feature of cubature does not allow the evaluation of multiple integration limits at once. Instead, it allows the integrator that is part of the core of `cubature` to evaluate multiple integration points at once, thus making it faster if your numerical library supports vectorisation (which is the case of NumPy).

In your example, I tested the answer provided in #33  and the vectorized version takes about 60% of the time of the non-vectorized one. To achieve the variable integration limits as you propose, you need to do something like:

```
from cubature import cubature
import numpy as np
import scipy.stats as ss

class Kelly:
    def __init__(self):
        self.theo = 0.556
        self.odds = 1.952
        self.kelly = self.calculate_kelly()
        self.x = 100
        self.n = 180
        self.a = 50
        self.b = 50
        self.i = 0.001
        self.f = np.arange(0, 1 + self.i, self.i)
        self.c1 = 1
        self.c2 = 2
        self.k = 1.5

    def calculate_kelly(self) -> float:
        return (self.theo * self.odds - 1) / (self.odds - 1)

    def calculate_loss_function_vec(self, p):
        f = self.f.flatten()
        p = p[:, 0]
        loss_function = np.where(p[:, None] < (f * self.odds - 1) + 1 / self.odds,
                                 (self.c1 + self.c2) * abs(f - self.kelly) ** self.k, 0)
        return loss_function * ss.beta.pdf(p, self.a + self.x, self.n - self.x + self.b)[:, None]

    def integrate_vec(self, xmin, xmax):
        vals, errors = cubature(func=self.calculate_loss_function_vec,
                                ndim=1,
                                fdim=len(self.f),
                                xmin=xmin,
                                xmax=xmax,
                                vectorized=True)
        return vals, errors


kelly = Kelly()
vals_all = []
xmax_all = (kelly.f * (kelly.odds - 1) + 1 / kelly.odds)
for xmax in xmax_all:
    vals, errors = kelly.integrate_vec(xmin=[0], xmax=[xmax])
    vals_all.append(vals)
```

**Comment by HJA24** [comment]
Thank you for taking a look again @saullocastro! You mentioned: "_The vectorized feature of cubature does not allow the evaluation of multiple integration limits at once."_. This refers to my input of argument `xmax`  in `cubature()` right?

`xmax=np.array([self.f * (self.odds - 1) + 1 / self.odds])`

Is this the case for this particular Python wrapper of _cubature_? Because when I run the following R script the integration is done relatively easy / fast?

```
library(cubature)
G <- function(f, x, n, theta, a, b, c1, c2, k){
    l_3_1 <- function(p){
        kelly <- ifelse(p > (1/theta), (p * theta - 1) / (theta - 1), 0)
        loss_func = ifelse(p < (f * (theta - 1) + 1)/theta,
                           yes = (c1 + c2) * (abs(f-kelly)^k),                            
                           no = 0)
        return(loss_func*dbeta(p, a + x, n - x + b))
        }
        
    l_3_2 <- function(p){
        kelly <- ifelse(p > (1/theta), (p * theta - 1) / (theta - 1), 0)
        loss_func = ifelse(p >= (f * (theta - 1) + 1)/theta,
                           yes = c2 * (abs(f-kelly)^k),
                           no = 0)

        return(loss_func*dbeta(p, a + x, n - x + b))
    }

    # Integrate the loss function
    part_1 <- adaptIntegrate(l_3_1, lowerLimit = 0, upperLimit = (f * (theta-1) + 1)/theta,
                             tol = 1e-10)
    part_2 <- adaptIntegrate(l_3_2, lowerLimit = (f * (theta-1) + 1)/theta , upperLimit = 1,
                             tol = 1e-10)

    # Return the sum of the two integrals
    return(part_1$integral + part_2$integral )
 }

f3 <- function(x, n, theta, a, b, c1, c2, k){
    i <- 0.001
    # f is a vector from 0 to 1
    f <- seq(0, 1, i)

    # G_ of_f contains the results of the function of G evalutated at all the f values
    G_of_f <- lapply(f, G, x = x, n = n, theta = theta, a = a, b = b, c1 = c1, c2 = c2, k = k)

    # Find the index of the minimum value and the corresponding f
    Modified_Kelly_Value <- which.min(G_of_f) * i - i

    return(Modified_Kelly_Value)
} 
 
x = 100
theta = 1.952
n = 180 
a =  b = 50
c1 = 1
c2 = 1
k = 1.5
modified_kelly = f3(x=x, n=n, theta=theta, a=a, b=b, c1=c1, c2=c2, k=k)
```

Looking at the [documentation ](https://www.rdocumentation.org/packages/cubature/versions/1.0/topics/adaptIntegrate) of `adaptintegrate`, can I assume that `cubature.cubature()` is the equivalent? If so, why can't they both do the desired integration? 
Thank you for your time again!

**Comment by saullocastro** [comment]
@HJA24 I could confirm today that there was a bug in the shape checks for `xmin` and `xmax` that were preventing the wrapper to use the full vectorized capability.

**Comment by HJA24** [comment]
Okay, glad I had a part in detecting it. Should be an easy fix right?

**Comment by saullocastro** [comment]
@HJA24  it is fixed in Version 0.16.0.

You can then test with 

```python

from cubature import cubature
import numpy as np
import scipy.stats as ss


class Kelly:
    def __init__(self):
        self.theo = 0.556
        self.odds = 1.952
        self.kelly = self.calculate_kelly()
        self.x = 100
        self.n = 180
        self.a = 50
        self.b = 50
        self.i = 0.001
        self.f = np.arange(0, 1 + self.i, self.i)
        self.c1 = 1
        self.c2 = 2
        self.k = 1.5

    def calculate_kelly(self) -> float:
        return (self.theo * self.odds - 1) / (self.odds - 1)

    def calculate_loss_function_vec(self, p):
        f = self.f.flatten()
        p = p[:, 0]
        loss_function = np.where(p[:, None] < (f * self.odds - 1) + 1 / self.odds,
                                 (self.c1 + self.c2) * abs(f - self.kelly) ** self.k, 0)
        return loss_function * ss.beta.pdf(p, self.a + self.x, self.n - self.x + self.b)[:, None]

    def integrate_vec(self):
        xmax = np.array([self.f * (self.odds - 1) + 1 / self.odds]).flatten()
        print(xmax.shape)
        vals, errors = cubature(func=self.calculate_loss_function_vec,
                                ndim=1,
                                fdim=len(self.f),
                                xmin=np.zeros(len(self.f)),
                                xmax=xmax,
                                vectorized=True)
        return vals, errors


kelly = Kelly()
vals, errors = kelly.integrate_vec()
print(vals, errors)

```

**Comment by saullocastro** [comment]
@HJA24 please check it and close the issue if everything is correct...
